src/interfaces/ipc.h

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INTERFACES_IPC_H
+#define BITCOIN_INTERFACES_IPC_H
+
+#include <fs.h>
+#include <interfaces/init.h>
+
+#include <memory>
+
+namespace interfaces {
+//! IPC process interface for spawning bitcoin processes and serving requests
+//! in processes that have been spawned.
+//!
+//! There will different implementations of this interface depending on the
+//! platform (e.g. unix, windows).
+class IpcProcess
+{
+public:
+    virtual ~IpcProcess() = default;
+
+    //! Spawn process and return socket file descriptor for communicating with
+    //! it.
+    virtual int spawn(const std::string& new_exe_name, int& pid) = 0;
+
+    //! Wait for spawned process to exit and return exit code.
+    virtual int wait(int pid) = 0;
+
+    //! Serve request if current process is a spawned subprocess. Blocks until
+    //! socket for communicating with the parent process is disconnected.
+    virtual bool serve(int& exit_status) = 0;
+};
+
+//! IPC protocol interface for calling IPC methods over sockets.
+//!
+//! There may be different implementations of this interface for different IPC
+//! protocols (e.g. Cap'n Proto, gRPC, JSON-RPC, or custom protocols).
+//!
+//! An implementation of this interface needs to provide an `interface::Init`
+//! object that translates method calls into requests sent over a socket, and it
+//! needs to implement a handler that translates requests received over a socket
+//! into method calls on a provided `interface::Init` object.
+class IpcProtocol
+{
+public:
+    virtual ~IpcProtocol() = default;
+
+    //! Return Init interface that forwards requests over given socket descriptor.
+    //! Socket communication is handled on a background thread.
+    //!
+    //! @note It could be potentially useful in the future to add
+    //! std::function<void()> on_disconnect callback argument here. But there
+    //! isn't an immediate need, because the protocol implementation can clean
+    //! up its own state (calling ProxyServer destructors, etc) on disconnect,
+    //! and any client calls will just throw IpcException errors after a
+    //! disconnect.
+    virtual std::unique_ptr<Init> connect(int fd) = 0;
+
+    //! Handle requests on provided socket descriptor. Socket communication is
+    //! handled on the current thread. This blocks until the client closes the socket.
+    //!
+    //! @note: If this method is called, it needs be called before connect()
+    //! because for ease of implementation it's inflexible and always runs the
+    //! event loop in the foreground thread. It can share its event loop with
+    //! connect() but can't share an event loop that was created by connect().
+    //! This isn't really a problem because serve() is only called by spawned
```

Comment:
643d602

Maybe the connection flow between parent and child process could be documented, what should be the order between parent serve, parent connect, child serve, child connect ? in `echoipc` demonstration, I don't see `serve` being called and thus that's a bit confusing with regards to documentation.

Reply:
> [643d602](https://github.com/bitcoin/bitcoin/commit/643d602b3fcb3eac22a2962cff1a0b7019bfebb0)
> 
> Maybe the connection flow between parent and child process could be documented, what should be the order between parent serve, parent connect, child serve, child connect ? in `echoipc` demonstration, I don't see `serve` being called and thus that's a bit confusing with regards to documentation.

Thanks for the review! I'll address the other comments, but just in case there is any question about `serve()` it's called in `main()` here in the `bitcoin-node` executable:

https://github.com/ryanofsky/bitcoin/blob/pr/ipc-echo.4/src/bitcoind.cpp#L178-L185

and here in the `bitcoin-wallet` executable:

https://github.com/ryanofsky/bitcoin/blob/pr/ipc.113/src/bitcoin-wallet.cpp#L80-L88

If we wanted to have a separate `bitcoin-echo` executable, it could have a `main()` function just like those examples. I didn't want to have a separate `bitcoin-echo` executable though, just because I thought it'd be confusing to have a kind of pointless `bitcoin-echo` program installed along with `bitcoin-gui` `bitcoin-node` and `bitcoin-wallet`, so I just put the echo functionality inside `bitcoin-node`.